core/src/test/scala/unit/kafka/log/LogCleanerTest.scala

```diff
@@ -270,6 +270,100 @@ class LogCleanerTest {
+    appendProducer1(Seq(2, 3))
+    log.appendAsLeader(abortMarker(producerId1, producerEpoch), leaderEpoch = 0, origin = AppendOrigin.Replication)
+    appendProducer2(Seq(4, 5))
+    appendProducer1(Seq(5, 6))
+    log.appendAsLeader(commitMarker(producerId1, producerEpoch), leaderEpoch = 0, origin = AppendOrigin.Replication)
+    log.appendAsLeader(abortMarker(producerId2, producerEpoch), leaderEpoch = 0, origin = AppendOrigin.Replication)
+    appendProducer2(Seq(6, 7))
+    log.appendAsLeader(commitMarker(producerId2, producerEpoch), leaderEpoch = 0, origin = AppendOrigin.Replication)
+    log.roll()
+
+    val expectedAbortedTxns = List(
+      new AbortedTxn(producerId=producerId1, firstOffset=8, lastOffset=10, lastStableOffset=11),
+      new AbortedTxn(producerId=producerId2, firstOffset=11, lastOffset=16, lastStableOffset=17)
+    )
+
+    assertAllTransactionsComplete(log)
+    assertAllAbortedTxns(expectedAbortedTxns, log)
+
+    var dirtyOffset = 0L
+    def cleanSegments(): Unit = {
+      val offsetMap = new FakeOffsetMap(slots = offsetMapSlots)
+      val segments = log.logSegments(0, log.activeSegment.baseOffset).toSeq
+      val stats = new CleanerStats(time)
+      cleaner.buildOffsetMap(log, dirtyOffset, log.activeSegment.baseOffset, offsetMap, stats)
+      cleaner.cleanSegments(log, segments, offsetMap, time.milliseconds(), stats, new CleanedTransactionMetadata, Long.MaxValue)
+      dirtyOffset = offsetMap.latestOffset + 1
+    }
+
+    // On the first pass, we should see the data from the aborted transactions deleted,
+    // but the markers should remain until the deletion retention time has passed.
+    cleanSegments()
+    assertEquals(4L, dirtyOffset)
+    assertEquals(List(0, 2, 4, 6, 7, 10, 13, 15, 16, 17, 19), batchBaseOffsetsInLog(log))
+    assertEquals(List(0, 2, 3, 4, 5, 6, 7, 10, 13, 14, 15, 16, 17, 18, 19), offsetsInLog(log))
+    assertAllTransactionsComplete(log)
+    assertAllAbortedTxns(expectedAbortedTxns, log)
+
+    // On the second pass, no data from the aborted transactions remains. The markers
+    // still cannot be removed from the log due to the retention time, but we do not
+    // need to record them in the transaction index since they are empty.
+    cleanSegments()
+    assertEquals(14, dirtyOffset)
+    assertEquals(List(0, 2, 4, 6, 7, 10, 13, 15, 16, 17, 19), batchBaseOffsetsInLog(log))
+    assertEquals(List(0, 2, 4, 5, 6, 7, 10, 13, 14, 15, 16, 17, 18, 19), offsetsInLog(log))
+    assertAllTransactionsComplete(log)
+    assertAllAbortedTxns(List(), log)
+
+    // On the last pass, wait for the retention time to expire. The abort markers
+    // (offsets 10 and 16) should be deleted.
+    time.sleep(deleteRetentionMs * 2)
+    cleanSegments()
+    assertEquals(20L, dirtyOffset)
+    assertEquals(List(0, 2, 4, 6, 7, 13, 15, 17, 19), batchBaseOffsetsInLog(log))
+    assertEquals(List(0, 2, 4, 5, 6, 7, 13, 15, 17, 18, 19), offsetsInLog(log))
```

Comment:
I am wondering why offset 16 (abort for p2) is removed in this round of cleaning. This is the first time that we clean offset 16. I was expecting this round only sets the delete horizon for the abort marker.

Reply:
I guess it's a little surprising, but the data from aborted transactions gets deleted even from the uncleaned portion of the log when we are building the new segment. In this test, that means the deletion horizon can be set in the second pass. I debating changing this behavior, but I couldn't come up with a strong reason to do so given that we have to do the work to build the transaction index anyway.

Reply:
Thanks for the explanation. Make sense to me.